<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Reproducing a single-cell proteomics data analysis using QFeatures and scp • QFeaturesScpWorkshop2021</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Reproducing a single-cell proteomics data analysis using QFeatures and scp">
<meta property="og:description" content="Recent advances in sample preparation, processing and mass
  spectrometry (MS) have enabled the emergence of MS-based single-cell
  proteomics (SCP). `QFeatures` along with its extension `scp` allow for
  standardized analysis of SCP data. `QFeatures` is based on the
  `SummarizedExperiment` and `MultiAssayExperiment` Bioconductor classes.
  Assays in a `QFeatures` object have a hierarchical relation: proteins
  are composed of peptides, themselves produced by spectra. Throughout
  the aggregation and processing of these data, the relations between
  assays are tracked and recorded. `QFeatures` is further able to track
  those relations across hundreds of acquisition batches that are
  usually required for SCP experiments. Users can therefore easily
  navigate across spectra, peptide and protein quantitative data within
  or between acquisition batches. In this workshop, we will reproduce an
  existing SCP data analysis on a subset of the original dataset. We
  will demonstrate how to import data containing multiple acquisition
  batches as `QFeatures` objects, how to process and analyse data in
  `QFeatures` objects, and how to interpret the results. Along the road,
  we will highlight the current limitation in the field when it comes to
  analysing SCP data, namely showing the presence of important batch
  effects and data missingness. Some familiarity with Bioconductor data
  analysis, in particular the `SummarizedExperiment` class, and basic
  knowledge about single-cell analyses are recommended to follow this
  workshop.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">QFeaturesScpWorkshop2021</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/v01-QFeatures.html">The QFeatures class</a>
    </li>
    <li>
      <a href="articles/v02-scp.html">Using scp for single-cell proteomics analyses</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lgatto/QFeaturesScpWorkshop2021">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="single-cell-proteomics-data-analysis-using-qfeatures-and-scp" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#single-cell-proteomics-data-analysis-using-qfeatures-and-scp" class="anchor"></a>Single-cell proteomics data analysis using QFeatures and scp</h1></div>
<p>Authors: Laurent Gatto and Christophe Vanderaa.</p>
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<div id="description" class="section level3">
<h3 class="hasAnchor">
<a href="#description" class="anchor"></a>Description</h3>
<p>Mass spectrometry (MS)-based single-cell proteomics (SCP) is an emerging field that requires a dedicated computational environment. <code>QFeatures</code> along with its extension <code>scp</code> allow for standardized analysis of SCP data. The workshop will start by introducing the <code>QFeatures</code> class and its functions to perform generic proteomics data analysis. We will then move to SCP and present how <code>scp</code> extends <code>QFeatures</code> to single-cell applications. The remainder of the workshop will be a hands-on session where attendees will be guided through the reproduction of a real-life analysis of published SCP data. Along the reproduction exercise, we will point out to current challenges that still need to be tackled computationally. This workshop is meant for inexperienced users that want to learn how to perform current state-of-the-art analysis of SCP data as well as experienced developers interested in contributing to an emerging and exciting single-cell technology.</p>
<p>This workshop is provided as two vignettes. The <a href="https://lgatto.github.io/QFeaturesScpWorkshop2021/articles/v01-QFeatures.html">first vignette</a> provides a general introduction to the <code>QFeatures</code> class in the general context of MS-based proteomics data manipulation. The <a href="https://lgatto.github.io/QFeaturesScpWorkshop2021/articles/v02-scp.html">second vignette</a> focuses on single-cell application and introduces the <code>scp</code> package as an extension of <code>QFeatures</code>. This second vignette also provides an exercise that gives the attendee the opportunity to apply the learned concepts to reproduce a published analysis on a subset of a real data set.</p>
</div>
<div id="pre-requisites" class="section level3">
<h3 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h3>
<ul>
<li>Basic knowledge of R syntax</li>
<li>Familiarity with the <code>SummarizedExperiment</code> class</li>
<li>Familiarity with the <code>tidyverse</code> ecosystem</li>
</ul>
<p>We recommend reading the paper that has published the SCP analysis that will be reproduced in this workshop:</p>
<p>Specht, Harrison, Edward Emmott, Aleksandra A. Petelski, R. Gray Huffman, David H. Perlman, Marco Serra, Peter Kharchenko, Antonius Koller, and Nikolai Slavov. 2021. “Single-Cell Proteomic and Transcriptomic Analysis of Macrophage Heterogeneity Using SCoPE2.” Genome Biology 22 (1): 50. <a href="http://dx.doi.org/10.1186/s13059-021-02267-5">link to article</a>, <a href="https://www.biorxiv.org/content/10.1101/665307v5">link to preprint</a></p>
</div>
<div id="r--bioconductor-packages-used" class="section level3">
<h3 class="hasAnchor">
<a href="#r--bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h3>
<p><code>QFeatures</code>, <code>scp</code>, <code>scpdata</code>, <code>MultiAssayExperiment</code>, <code>SingleCellExperiment</code></p>
</div>
<div id="time-outline" class="section level3">
<h3 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h3>
<p>This 90 min workshop will be split in three parts:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introduction to the <code>QFeatures</code> class and functions</td>
<td>25m</td>
</tr>
<tr class="even">
<td>Introduction to SCP and the <code>scp</code> package</td>
<td>20m</td>
</tr>
<tr class="odd">
<td>Reproducing SCoPE2: a published SCP data analysis</td>
<td>45m</td>
</tr>
</tbody>
</table>
</div>
<div id="learning-goals" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-goals" class="anchor"></a>Learning goals</h3>
<ul>
<li>Highlight the common steps to perform MS-based proteomics data analysis</li>
<li>Understand the specificities when it comes to single-cell data</li>
<li>Identify the current challenges in the SCP data analysis</li>
</ul>
</div>
<div id="learning-objectives" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-objectives" class="anchor"></a>Learning objectives</h3>
<ul>
<li>Use <code>QFeatures</code> and <code>scp</code> to perform a real-life analysis of SCP data</li>
<li>Integrate the workflow with other tools such as the <code>tidyverse</code> ecosystem for efficient data visualization or the <code>SingleCellExperiment</code> related tools for extending the current analysis workflow.</li>
</ul>
</div>
</div>
<div id="running-the-workshop" class="section level2">
<h2 class="hasAnchor">
<a href="#running-the-workshop" class="anchor"></a>Running the workshop</h2>
<p>You can install all necessary packages to run the workshop locally by installing the following packages:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Bioconductor packages</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"QFeatures"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"scp"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"scpdata"</span><span class="op">)</span></code></pre></div>
<p>The <a href="https://hub.docker.com/repository/docker/lgatto/qfeaturesscpworkshop2021">lgatto/qfeaturesscpworkshop2021</a> docker container has all packages necessary for running the workshop vignettes.</p>
<p>The container can be downloaded and run with</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">-e</span> PASSWORD=bioc <span class="at">-p</span> 8787:8787 lgatto/qfeaturesscpworkshop2021:latest</span></code></pre></div>
<p>(you can choose any password, not only <code>bioc</code>, above)</p>
<p>Once running, navigate to <a href="https://localhost:8787/" class="uri">https://localhost:8787/</a> and then login with user <code>rstudio</code> and password <code>bioc</code>.</p>
<p>During the <a href="https://bioc2021.bioconductor.org/">Bioc2021</a> conference, the workshop can be <a href="http://app.orchestra.cancerdatasci.org/">run in the cloud</a>.</p>
<p>The content of this workshop is provided under a <a href="https://creativecommons.org/licenses/by-sa/2.0/">CC-BY ShareAlike</a> license.</p>
<pre><code>
To cite package 'QFeaturesScpWorkshop2021' in publications use:

  Laurent Gatto and Christophe Vanderaa (NA). QFeaturesScpWorkshop2021:
  Reproducing a single-cell proteomics data analysis using QFeatures
  and scp. R package version 0.1.0.
  https://github.com/lgatto/QFeaturesScpWorkshop2021

A BibTeX entry for LaTeX users is

  @Manual{,
    title = {QFeaturesScpWorkshop2021: Reproducing a single-cell proteomics data analysis using QFeatures and scp.},
    author = {Laurent Gatto and Christophe Vanderaa},
    note = {R package version 0.1.0},
    url = {https://github.com/lgatto/QFeaturesScpWorkshop2021},
  }</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/lgatto/QFeaturesScpWorkshop2021/">https://​github.com/​lgatto/​QFeaturesScpWorkshop2021/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/lgatto/QFeaturesScpWorkshop2021/issues/">https://​github.com/​lgatto/​QFeaturesScpWorkshop2021/​issues/​</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/Artistic-2.0">Artistic-2.0</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Laurent Gatto <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-1520-2268" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Christophe Vanderaa <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-7443-5427" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Laurent Gatto, Christophe Vanderaa.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
