<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using scp for single-cell proteomics analyses • QFeaturesScpWorkshop2021</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using scp for single-cell proteomics analyses">
<meta property="og:description" content="QFeaturesScpWorkshop2021">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">QFeaturesScpWorkshop2021</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/v01-QFeatures.html">The QFeatures class</a>
    </li>
    <li>
      <a href="../articles/v02-scp.html">Using scp for single-cell proteomics analyses</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lgatto/QFeaturesScpWorkshop2021">
    <span class="fas fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="v02-scp_files/header-attrs-2.8/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using scp for single-cell proteomics analyses</h1>
                        <h4 class="author">Laurent Gatto and Christophe Vanderaa</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/lgatto/QFeaturesScpWorkshop2021/blob/master/vignettes/v02-scp.Rmd"><code>vignettes/v02-scp.Rmd</code></a></small>
      <div class="hidden name"><code>v02-scp.Rmd</code></div>

    </div>

    
    
<p><strong>Last modified:</strong> 2021-05-28 17:56:32<br><strong>Compiled</strong>: Fri May 28 20:33:50 2021</p>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Mass spectrometry (MS)-based single-cell proteomics (SCP) is emerging thanks to several recent technological advances in sample preparation, liquid chromatography (LC) and MS (see <span class="citation">Kelly (2020)</span> for a comprehensive review). The improvements tackle the issues encountered when dealing with small sample amounts and focus on:</p>
<ul>
<li>Reducing sample loss</li>
<li>Increasing sensitivity and quantification accuracy</li>
<li>Increasing acquisition throughput</li>
</ul>
<p>Two main strategies have currently been developed. On the one hand, label-free protocols acquire one single cell per MS run leading to accurate quantification but low sensitivity and throughput. On the other hand, label-based protocols multiplex several single-cell samples in one MS run leading to higher throughput (1000 cells per week). Another advantage of label-based protocols is the inclusion of a carrier sample, that is a sample containing between tens to hundreds of cells. Including a carrier increases sensitivity thanks to increased sample material, but at the cost of decreased quantification accuracy due to chemical noise (linked to sample labelling) and competition between the single-cell samples and the carrier sample during MS acquisition.</p>
</div>
<div id="the-scp-data-framework" class="section level1">
<h1 class="hasAnchor">
<a href="#the-scp-data-framework" class="anchor"></a>The <code>scp</code> data framework</h1>
<p>SCP data is very similar to bulk proteomics data with the exception that the PSM data may be composed of tens to hundreds of separate acquisition runs. The <code>QFeatures</code> class is able to store this acquisition structure by considering each MS run as a separate assay. The assays can be subjected to downstream analysis for single-cell data such as dimensions reduction, clustering, differential abundance analysis,… Therefore, the assays are stored as <code>SingleCellExperiment</code> objects (<span class="citation">Lun and Risso (2020)</span>) to create a direct interface to existing Bioconductor packages to perform such downstream analyses. The links between related features across different assays are also stored to facilitate manipulation and visualization of of PSM, peptide and protein data as we go along with the processing workflow.</p>
<div class="figure">
<img src="figs/SCP_framework.png" alt="Conceptual overview of a `QFeatures` object containing SCP data. Each assay is stored as a `SingleCellExperiment` object." width="342"><p class="caption">
Conceptual overview of a <code>QFeatures</code> object containing SCP data. Each assay is stored as a <code>SingleCellExperiment</code> object.
</p>
</div>
</div>
<div id="scp-data" class="section level1">
<h1 class="hasAnchor">
<a href="#scp-data" class="anchor"></a>SCP data</h1>
<p>There are two input tables required for starting an analysis with <code>scp</code>:</p>
<ol style="list-style-type: decimal">
<li>The feature data</li>
<li>The sample data</li>
</ol>
<div id="feature-data" class="section level2">
<h2 class="hasAnchor">
<a href="#feature-data" class="anchor"></a>Feature data</h2>
<p>The feature data are generated after the identification and quantification of the MS spectra by a pre-processing software such as MaxQuant, ProteomeDiscoverer or MSFragger (the <a href="https://en.wikipedia.org/wiki/List_of_mass_spectrometry_software">list</a> of available software is actually much longer). We will here use as an example a data table that has been generated by MaxQuant. The table is available from the <code>scp</code> package and is called <code>mqScpData</code> (for MaxQuant generated SCP data).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://UCLouvain-CBIO.github.io/scp">scp</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"mqScpData"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">mqScpData</span><span class="op">)</span>
<span class="co">## [1] 1361  149</span></code></pre></div>
<p>In this toy example, there are 1361 rows corresponding to features (quantified PSMs) and 149 columns corresponding to different data fields recorded by MaxQuant during the processing of the MS spectra. The columns can be divided into three categories:</p>
<ul>
<li>Columns holding feature quantifications</li>
<li>Columns holding feature metadata</li>
<li>Columns holding MS run metadata</li>
</ul>
<div id="feature-quantifications" class="section level3">
<h3 class="hasAnchor">
<a href="#feature-quantifications" class="anchor"></a>Feature quantifications</h3>
<p>The quantification data can be composed of one (in case of label-free acquisition) up to 16 columns (in case of TMT-16 multiplexing). The columns holding the quantification start with <code>Reporter.intensity.</code> followed by a number.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="op">(</span><span class="va">quantCols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"Reporter.intensity.\\d"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">mqScpData</span><span class="op">)</span>,
                  value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="co">##  [1] "Reporter.intensity.1"  "Reporter.intensity.2"  "Reporter.intensity.3" </span>
<span class="co">##  [4] "Reporter.intensity.4"  "Reporter.intensity.5"  "Reporter.intensity.6" </span>
<span class="co">##  [7] "Reporter.intensity.7"  "Reporter.intensity.8"  "Reporter.intensity.9" </span>
<span class="co">## [10] "Reporter.intensity.10" "Reporter.intensity.11" "Reporter.intensity.12"</span>
<span class="co">## [13] "Reporter.intensity.13" "Reporter.intensity.14" "Reporter.intensity.15"</span>
<span class="co">## [16] "Reporter.intensity.16"</span></code></pre></div>
<p>As you may notice, the example data was acquired using a TMT-16 protocol since we retrieve 16 quantification columns. Actually, some runs were acquired using a TMT-11 protocol (11 labels) but we will come back to this later.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">mqScpData</span><span class="op">[</span>, <span class="va">quantCols</span><span class="op">]</span><span class="op">)</span>
<span class="co">##   Reporter.intensity.1 Reporter.intensity.2 Reporter.intensity.3</span>
<span class="co">## 1                61251               501.71               3731.3</span>
<span class="co">## 2                58648              1099.80               2837.7</span>
<span class="co">## 3               150350              3705.00               9361.0</span>
<span class="co">## 4                27347               405.90               1525.2</span>
<span class="co">## 5                84035               583.09               4092.3</span>
<span class="co">## 6                44895               700.23               2283.0</span>
<span class="co">##   Reporter.intensity.4 Reporter.intensity.5 Reporter.intensity.6</span>
<span class="co">## 1              1643.30               871.84               981.87</span>
<span class="co">## 2               494.32               349.26              1030.50</span>
<span class="co">## 3                 0.00              1945.40              1188.60</span>
<span class="co">## 4                 0.00                 0.00               318.74</span>
<span class="co">## 5               530.13               718.13              2204.50</span>
<span class="co">## 6              1109.60                 0.00               675.79</span>
<span class="co">##   Reporter.intensity.7 Reporter.intensity.8 Reporter.intensity.9</span>
<span class="co">## 1              1200.10               939.06              1457.50</span>
<span class="co">## 2                 0.00              1214.10               800.58</span>
<span class="co">## 3              1574.00              2302.10              2176.10</span>
<span class="co">## 4                 0.00               519.81                 0.00</span>
<span class="co">## 5               960.51               453.77              1188.40</span>
<span class="co">## 6                 0.00               809.38               668.88</span>
<span class="co">##   Reporter.intensity.10 Reporter.intensity.11 Reporter.intensity.12</span>
<span class="co">## 1               1329.80                981.83                    NA</span>
<span class="co">## 2                807.79                391.38                    NA</span>
<span class="co">## 3               1399.50               1307.50                2192.4</span>
<span class="co">## 4                507.23                370.79                    NA</span>
<span class="co">## 5                740.99                  0.00                    NA</span>
<span class="co">## 6               1467.50                901.38                    NA</span>
<span class="co">##   Reporter.intensity.13 Reporter.intensity.14 Reporter.intensity.15</span>
<span class="co">## 1                    NA                    NA                    NA</span>
<span class="co">## 2                    NA                    NA                    NA</span>
<span class="co">## 3                1791.4                1727.5                2157.3</span>
<span class="co">## 4                    NA                    NA                    NA</span>
<span class="co">## 5                    NA                    NA                    NA</span>
<span class="co">## 6                    NA                    NA                    NA</span>
<span class="co">##   Reporter.intensity.16</span>
<span class="co">## 1                    NA</span>
<span class="co">## 2                    NA</span>
<span class="co">## 3                  1398</span>
<span class="co">## 4                    NA</span>
<span class="co">## 5                    NA</span>
<span class="co">## 6                    NA</span></code></pre></div>
</div>
<div id="feature-metadata" class="section level3">
<h3 class="hasAnchor">
<a href="#feature-metadata" class="anchor"></a>Feature metadata</h3>
<p>Most columns in the <code>mqScpData</code> table contain information used or generated during the identification of the MS spectra. For instance, you may find the charge of the parent ion, the score and probability of a correct match between the MS spectrum and a peptide sequence, the sequence of the best matching peptide, its length, its modifications, the retention time of the peptide on the LC, the protein(s) the peptide originates from and much more.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">mqScpData</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Charge"</span>, <span class="st">"Score"</span>, <span class="st">"PEP"</span>, <span class="st">"Sequence"</span>, <span class="st">"Length"</span>,
                   <span class="st">"Retention.time"</span>, <span class="st">"Proteins"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="co">##   Charge  Score        PEP    Sequence Length Retention.time</span>
<span class="co">## 1      2 41.029 5.2636e-04   ATNFLAHEK      9         65.781</span>
<span class="co">## 2      2 44.349 5.8789e-04   ATNFLAHEK      9         63.787</span>
<span class="co">## 3      2 51.066 4.0315e-24 SHTILLVQPTK     11         71.884</span>
<span class="co">## 4      2 63.816 4.7622e-06 SHTILLVQPTK     11         68.633</span>
<span class="co">## 5      2 74.464 6.8709e-09 SHTILLVQPTK     11         71.946</span>
<span class="co">## 6      2 41.502 5.3705e-02     SLVIPEK      7         76.204</span>
<span class="co">##               Proteins</span>
<span class="co">## 1 sp|P29692|EF1D_HUMAN</span>
<span class="co">## 2 sp|P29692|EF1D_HUMAN</span>
<span class="co">## 3  sp|P84090|ERH_HUMAN</span>
<span class="co">## 4  sp|P84090|ERH_HUMAN</span>
<span class="co">## 5  sp|P84090|ERH_HUMAN</span>
<span class="co">## 6 sp|P62269|RS18_HUMAN</span></code></pre></div>
</div>
<div id="ms-run-metadata" class="section level3">
<h3 class="hasAnchor">
<a href="#ms-run-metadata" class="anchor"></a>MS run metadata</h3>
<p>This type of metadata is related to the MS instrument. In MaxQuant, only the file name generated by MS instrument is stored. There is one file for each MS run, hence the file name can be used as a batch identifier.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">mqScpData</span><span class="op">$</span><span class="va">Raw.file</span><span class="op">)</span>
<span class="co">## [1] "190321S_LCA10_X_FP97AG"        "190222S_LCA9_X_FP94BM"        </span>
<span class="co">## [3] "190914S_LCB3_X_16plex_Set_21"  "190321S_LCA10_X_FP97_blank_01"</span></code></pre></div>
<p>The 1361 PSM were found across 4 MS runs.</p>
</div>
</div>
<div id="sample-data" class="section level2">
<h2 class="hasAnchor">
<a href="#sample-data" class="anchor"></a>Sample data</h2>
<p>Next to the quantification data and the feature data, sample metadata contains the experimental design generated by the researcher. The rows of sample metadata correspond to a sample in the experiment and the columns correspond to the available information about the sample. We will here use the second example table provided in <code>scp</code>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"sampleAnnotation"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">sampleAnnotation</span><span class="op">)</span>
<span class="co">##                Raw.file              Channel SampleType lcbatch sortday digest</span>
<span class="co">## 1 190222S_LCA9_X_FP94BM Reporter.intensity.1    Carrier    LCA9      s8      N</span>
<span class="co">## 2 190222S_LCA9_X_FP94BM Reporter.intensity.2  Reference    LCA9      s8      N</span>
<span class="co">## 3 190222S_LCA9_X_FP94BM Reporter.intensity.3     Unused    LCA9      s8      N</span>
<span class="co">## 4 190222S_LCA9_X_FP94BM Reporter.intensity.4   Monocyte    LCA9      s8      N</span>
<span class="co">## 5 190222S_LCA9_X_FP94BM Reporter.intensity.5      Blank    LCA9      s8      N</span>
<span class="co">## 6 190222S_LCA9_X_FP94BM Reporter.intensity.6   Monocyte    LCA9      s8      N</span></code></pre></div>
<p>This table may contain any information about the samples. For example, useful information could be the type of sample that is analysed, an <em>apriori</em> known phenotype, the MS batch, the acquisition date, MS settings used to acquire the sample, the LC batch, the sample preparation batch, … <code>scp</code> <strong>requires</strong> 2 specific fields in the sample metadata:</p>
<ol style="list-style-type: decimal">
<li>One column containing the MS run names (<code>Raw.file</code> in this case). It must have the same name as the name of the column containing the MS run names in the quantification table. This will allow <code>scp</code> to correctly match and split data that were acquired separately. This is illustrated below by the linked circles.</li>
<li>One column that tells <code>scp</code> which column in the feature data holds the quantification of the corresponding sample. This is illustrated below by the arrow.</li>
</ol>
<div class="figure">
<img src="figs/readSCP_relation.png" alt="Linking the sample data to the feature data" width="196"><p class="caption">
Linking the sample data to the feature data
</p>
</div>
<p>As a note, the example sample data contains 5 different types of samples (<code>SampleType</code>):</p>
<ul>
<li>The carrier channels (<code>Carrier</code>) contain 200 cell equivalents and are meant to boost the peptide identification rate.</li>
<li>The normalization channels (<code>Reference</code>) contain 5 cell equivalents and are used to partially correct for between-run variation.</li>
<li>The unused channels (<code>Unused</code>) are channels that are left empty due to isotopic cross-contamination by the carrier channel.</li>
<li>The blank channels (<code>Blank</code>) contain samples that do not contain any cell but are processed as single-cell samples.</li>
<li>The single-cell sample channels contain the single-cell samples of interest, that are macrophage (<code>Macrophage</code>) or monocyte (<code>Monocyte</code>).</li>
</ul>
</div>
<div id="readscp" class="section level2">
<h2 class="hasAnchor">
<a href="#readscp" class="anchor"></a><code>readSCP</code>
</h2>
<p><code>readSCP</code> is the function that converts the sample and the feature data into a <code>QFeatures</code> object following the data structure described above, that is storing the data belonging to each MS batch in a separate <code>SingleCellExperiment</code> object. We therefore provide the feature data, the sample data to the function as well as the name of the column that holds the batch name in both tables and the name of the column in the sample data that points to the quantification columns in the feature data. We also add an (optional) suffix to the sample names to remind that samples were labelled using 16 different TMT. Finally, another optional argument is to remove empty samples (<code>removeEmptyCols = TRUE</code>). We do this because one run was acquired using TMT-16 and the three other runs were acquired using TMT-11, meaning that the quantification columns for labels 12 to 16 are empty for all except one batch.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/readSCP.html">readSCP</a></span><span class="op">(</span>featureData <span class="op">=</span> <span class="va">mqScpData</span>,
               colData <span class="op">=</span> <span class="va">sampleAnnotation</span>,
               batchCol <span class="op">=</span> <span class="st">"Raw.file"</span>,
               channelCol <span class="op">=</span> <span class="st">"Channel"</span>,
               suffix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"_TMT"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">16</span><span class="op">)</span>,
               removeEmptyCols <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>As prompted by the function, the feature data is converted to a <code>SingleCellExperiment</code> object then split according to the batch structure. This means that we should have as many assays than batches:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span>
<span class="co">## An instance of class QFeatures containing 4 assays:</span>
<span class="co">##  [1] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 395 rows and 11 columns </span>
<span class="co">##  [2] 190321S_LCA10_X_FP97_blank_01: SingleCellExperiment with 109 rows and 11 columns </span>
<span class="co">##  [3] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 487 rows and 11 columns </span>
<span class="co">##  [4] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 370 rows and 16 columns</span></code></pre></div>
<p>Indeed, we can see that the object returned by <code>readSCP</code> is a <code>QFeatures</code> object containing 4 <code>SingleCellExperiment</code> assays that have been named after the 4 MS batches. Each assay contains either 11 or 16 columns (samples) depending on the TMT labelling strategy and a variable number of rows (quantified PSMs). Each piece of information can easily be retrieved thanks to the <code>Qfeatures</code> architectures. As mentioned in the previous vignette, sample data is retrieved from the <code>colData</code>. Note that unique sample names were automatically generated by combining the batch name and the suffix provided to <code>readSCP</code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">)</span>
<span class="co">## DataFrame with 6 rows and 6 columns</span>
<span class="co">##                                 Raw.file       Channel  SampleType     lcbatch</span>
<span class="co">##                              &lt;character&gt;   &lt;character&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 190222S_LCA9_X_FP94BM_TMT1 190222S_LC... Reporter.i...     Carrier        LCA9</span>
<span class="co">## 190222S_LCA9_X_FP94BM_TMT2 190222S_LC... Reporter.i...   Reference        LCA9</span>
<span class="co">## 190222S_LCA9_X_FP94BM_TMT3 190222S_LC... Reporter.i...      Unused        LCA9</span>
<span class="co">## 190222S_LCA9_X_FP94BM_TMT4 190222S_LC... Reporter.i...    Monocyte        LCA9</span>
<span class="co">## 190222S_LCA9_X_FP94BM_TMT5 190222S_LC... Reporter.i...       Blank        LCA9</span>
<span class="co">## 190222S_LCA9_X_FP94BM_TMT6 190222S_LC... Reporter.i...    Monocyte        LCA9</span>
<span class="co">##                                sortday      digest</span>
<span class="co">##                            &lt;character&gt; &lt;character&gt;</span>
<span class="co">## 190222S_LCA9_X_FP94BM_TMT1          s8           N</span>
<span class="co">## 190222S_LCA9_X_FP94BM_TMT2          s8           N</span>
<span class="co">## 190222S_LCA9_X_FP94BM_TMT3          s8           N</span>
<span class="co">## 190222S_LCA9_X_FP94BM_TMT4          s8           N</span>
<span class="co">## 190222S_LCA9_X_FP94BM_TMT5          s8           N</span>
<span class="co">## 190222S_LCA9_X_FP94BM_TMT6          s8           N</span></code></pre></div>
<p>The feature metadata is retrieved from the <code>rowData</code>. Since the feature metadata is specific to each assay, we need to tell from which assay we want to get the <code>rowData</code>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">scp</span><span class="op">[[</span><span class="st">"190222S_LCA9_X_FP94BM"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">## DataFrame with 6 rows and 5 columns</span>
<span class="co">##                 uid      Sequence    Length Modifications Modified.sequence</span>
<span class="co">##         &lt;character&gt;   &lt;character&gt; &lt;integer&gt;   &lt;character&gt;       &lt;character&gt;</span>
<span class="co">## PSM2  _(Acetyl (...     ATNFLAHEK         9 Acetyl (Pr...     _(Acetyl (...</span>
<span class="co">## PSM4  _(Acetyl (... SHTILLVQPT...        11 Acetyl (Pr...     _(Acetyl (...</span>
<span class="co">## PSM6  _(Acetyl (...       SLVIPEK         7 Acetyl (Pr...     _(Acetyl (...</span>
<span class="co">## PSM9  _AAGLALK_ ...       AAGLALK         7    Unmodified         _AAGLALK_</span>
<span class="co">## PSM12 _AALSAGK_ ...       AALSAGK         7    Unmodified         _AALSAGK_</span>
<span class="co">## PSM15 _AAPEASGTP... AAPEASGTPS...        16    Unmodified     _AAPEASGTP...</span></code></pre></div>
<p>Finally, we can also retrieve the quantification matrix for an assay of interest:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">scp</span>, <span class="st">"190222S_LCA9_X_FP94BM"</span><span class="op">)</span><span class="op">)</span>
<span class="co">##       190222S_LCA9_X_FP94BM_TMT1 190222S_LCA9_X_FP94BM_TMT2</span>
<span class="co">## PSM2                     58648.0                    1099.80</span>
<span class="co">## PSM4                     27347.0                     405.90</span>
<span class="co">## PSM6                     44895.0                     700.23</span>
<span class="co">## PSM9                    122070.0                    1153.50</span>
<span class="co">## PSM12                    58605.0                     895.25</span>
<span class="co">## PSM15                     5006.5                     517.86</span>
<span class="co">##       190222S_LCA9_X_FP94BM_TMT3 190222S_LCA9_X_FP94BM_TMT4</span>
<span class="co">## PSM2                     2837.70                     494.32</span>
<span class="co">## PSM4                     1525.20                       0.00</span>
<span class="co">## PSM6                     2283.00                    1109.60</span>
<span class="co">## PSM9                     7361.90                    1732.30</span>
<span class="co">## PSM12                    2763.80                     867.82</span>
<span class="co">## PSM15                     446.19                     458.17</span>
<span class="co">##       190222S_LCA9_X_FP94BM_TMT5 190222S_LCA9_X_FP94BM_TMT6</span>
<span class="co">## PSM2                      349.26                    1030.50</span>
<span class="co">## PSM4                        0.00                     318.74</span>
<span class="co">## PSM6                        0.00                     675.79</span>
<span class="co">## PSM9                     1515.60                    2252.00</span>
<span class="co">## PSM12                    1050.30                    1268.70</span>
<span class="co">## PSM15                     467.90                     649.50</span>
<span class="co">##       190222S_LCA9_X_FP94BM_TMT7 190222S_LCA9_X_FP94BM_TMT8</span>
<span class="co">## PSM2                        0.00                    1214.10</span>
<span class="co">## PSM4                        0.00                     519.81</span>
<span class="co">## PSM6                        0.00                     809.38</span>
<span class="co">## PSM9                      444.48                    2343.80</span>
<span class="co">## PSM12                     532.30                    1073.10</span>
<span class="co">## PSM15                     259.84                     533.55</span>
<span class="co">##       190222S_LCA9_X_FP94BM_TMT9 190222S_LCA9_X_FP94BM_TMT10</span>
<span class="co">## PSM2                      800.58                      807.79</span>
<span class="co">## PSM4                        0.00                      507.23</span>
<span class="co">## PSM6                      668.88                     1467.50</span>
<span class="co">## PSM9                     3100.20                     1825.20</span>
<span class="co">## PSM12                     911.30                     1300.00</span>
<span class="co">## PSM15                     393.53                      463.26</span>
<span class="co">##       190222S_LCA9_X_FP94BM_TMT11</span>
<span class="co">## PSM2                       391.38</span>
<span class="co">## PSM4                       370.79</span>
<span class="co">## PSM6                       901.38</span>
<span class="co">## PSM9                      2372.50</span>
<span class="co">## PSM12                     1185.90</span>
<span class="co">## PSM15                      353.04</span></code></pre></div>
</div>
</div>
<div id="the-scp-package" class="section level1">
<h1 class="hasAnchor">
<a href="#the-scp-package" class="anchor"></a>The <code>scp</code> package</h1>
<p>The general workflow for processing SCP data is very similar to the workflow for bulk proteomics that we presented in the previous vignette. Therefore, the <code>QFeatures</code> package already contains most of the tools required for the processing of SCP data. The <code>scp</code> package implements the missing functions that are specifically designed for dealing with SCP data. We already used the <code>readSCP</code> function to format data from table to <code>Qfeatures</code> object. Here is a list of additional functions from <code>scp</code> that extend the <code>QFeatures</code> functions:</p>
<ul>
<li>
<code>aggregateFeaturesOverAssays</code>: extends <code>aggregateFeatures</code> to allow streamlined aggregation over multiple assays</li>
<li>
<code>computeSCR</code>: compute the sample over carrier ratio (SCR), a useful metric for feature QC</li>
<li>
<code>divideByReference</code>: divide columns by a reference column</li>
<li>
<code>medianCVperCell</code>: compute the median coefficient of variation (CV) per cell, a useful metric for single-cell QC</li>
<li>
<code>normalizeSCP</code>: extends <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html">QFeatures::normalize</a></code> to <code>SingleCellExperiment</code> objects</li>
<li>
<code>pep2qvalue</code>: compute q-values from posterior error probabilities</li>
<li>
<code>rowDataToDF</code>: extract the <code>rowData</code> of a <code>QFeatures</code> object to a <code>DataFrame</code>
</li>
</ul>
<p>You will be able to test those functions in the exercise later in this vignette.</p>
</div>
<div id="the-scpdata-package" class="section level1">
<h1 class="hasAnchor">
<a href="#the-scpdata-package" class="anchor"></a>The <code>scpdata</code> package</h1>
<p>Next to <code>scp</code>, we also developed <code>scpdata</code>, a data package that disseminates published SCP data sets formatted using the <code>scp</code> data structure. The package heavily relies on the <a href="https://bioconductor.org/packages/release/bioc/html/ExperimentHub.html"><code>ExperimentHub</code></a> infrastructure. This package is an ideal platform for data sharing and promotes for open and reproducible science in SCP, it facilitates the access for developers to SCP data to build and benchmark new methodologies and it facilitates the access for new users to data in the context of training and demonstration (like this workshop).</p>
<p>After loading the package, you can have a look at the available datasets by running:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scpdata</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/scpdata/man/scpdata.html">scpdata</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">## DataFrame with 12 rows and 15 columns</span>
<span class="co">##                title  dataprovider       species taxonomyid      genome</span>
<span class="co">##          &lt;character&gt;   &lt;character&gt;   &lt;character&gt;  &lt;integer&gt; &lt;character&gt;</span>
<span class="co">## EH3899 specht2019... SlavovLab ... Homo sapie...       9606          NA</span>
<span class="co">## EH3900 specht2019... SlavovLab ... Homo sapie...       9606          NA</span>
<span class="co">## EH3901 dou2019_ly...       MassIVE Homo sapie...       9606          NA</span>
<span class="co">## EH3902 dou2019_mo...       MassIVE Mus muscul...      10090          NA</span>
<span class="co">## EH3903 dou2019_bo...       MassIVE Mus muscul...      10090          NA</span>
<span class="co">## ...              ...           ...           ...        ...         ...</span>
<span class="co">## EH3906 zhu2018NC_...         PRIDE Homo sapie...       9606          NA</span>
<span class="co">## EH3907 zhu2018NC_...         PRIDE Homo sapie...       9606          NA</span>
<span class="co">## EH3908    cong2020AC         PRIDE Homo sapie...       9606          NA</span>
<span class="co">## EH3909     zhu2019EL         PRIDE Gallus gal...       9031          NA</span>
<span class="co">## EH6011 liang2020_...         PRIDE Homo sapie...       9606          NA</span>
<span class="co">##          description coordinate_1_based    maintainer rdatadateadded</span>
<span class="co">##          &lt;character&gt;          &lt;integer&gt;   &lt;character&gt;    &lt;character&gt;</span>
<span class="co">## EH3899 SCP expres...                  1 Christophe...     2020-11-05</span>
<span class="co">## EH3900 SCP expres...                  1 Christophe...     2020-11-05</span>
<span class="co">## EH3901 SCP expres...                  1 Christophe...     2020-11-05</span>
<span class="co">## EH3902 SCP expres...                  1 Christophe...     2020-11-05</span>
<span class="co">## EH3903 SCP expres...                  1 Christophe...     2020-11-05</span>
<span class="co">## ...              ...                ...           ...            ...</span>
<span class="co">## EH3906 Near SCP e...                  1 Christophe...     2020-11-05</span>
<span class="co">## EH3907 Near SCP e...                  1 Christophe...     2020-11-05</span>
<span class="co">## EH3908 SCP expres...                  1 Christophe...     2020-11-05</span>
<span class="co">## EH3909 SCP expres...                  1 Christophe...     2020-11-05</span>
<span class="co">## EH6011 Expression...                  1 Christophe...     2021-04-27</span>
<span class="co">##        preparerclass                                          tags  rdataclass</span>
<span class="co">##          &lt;character&gt;                                        &lt;list&gt; &lt;character&gt;</span>
<span class="co">## EH3899       scpdata Experiment...,Expression...,Experiment...,...   QFeatures</span>
<span class="co">## EH3900       scpdata Experiment...,Expression...,Experiment...,...   QFeatures</span>
<span class="co">## EH3901       scpdata Experiment...,Expression...,Experiment...,...   QFeatures</span>
<span class="co">## EH3902       scpdata Experiment...,Expression...,Experiment...,...   QFeatures</span>
<span class="co">## EH3903       scpdata Experiment...,Expression...,Experiment...,...   QFeatures</span>
<span class="co">## ...              ...                                           ...         ...</span>
<span class="co">## EH3906       scpdata Experiment...,Expression...,Experiment...,...   QFeatures</span>
<span class="co">## EH3907       scpdata Experiment...,Expression...,Experiment...,...   QFeatures</span>
<span class="co">## EH3908       scpdata Experiment...,Expression...,Experiment...,...   QFeatures</span>
<span class="co">## EH3909       scpdata Experiment...,Expression...,Experiment...,...   QFeatures</span>
<span class="co">## EH6011       scpdata      Expression...,MassSpectr...,Proteome,...   QFeatures</span>
<span class="co">##            rdatapath     sourceurl  sourcetype</span>
<span class="co">##          &lt;character&gt;   &lt;character&gt; &lt;character&gt;</span>
<span class="co">## EH3899 scpdata/sp... https://sc...         CSV</span>
<span class="co">## EH3900 scpdata/sp... https://sc...         CSV</span>
<span class="co">## EH3901 scpdata/do... ftp://mass...    XLS/XLSX</span>
<span class="co">## EH3902 scpdata/do... ftp://mass...    XLS/XLSX</span>
<span class="co">## EH3903 scpdata/do... ftp://mass...    XLS/XLSX</span>
<span class="co">## ...              ...           ...         ...</span>
<span class="co">## EH3906 scpdata/zh... ftp://ftp....         TXT</span>
<span class="co">## EH3907 scpdata/zh... ftp://ftp....         TXT</span>
<span class="co">## EH3908 scpdata/co... ftp://ftp....         TXT</span>
<span class="co">## EH3909 scpdata/zh... ftp://ftp....         TXT</span>
<span class="co">## EH6011 scpdata/li... ftp://ftp....         TXT</span></code></pre></div>
<p>For instance, loading the data sets published in <span class="citation">Zhu et al. (2019)</span> is as simple as calling the title of the data set:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/scpdata/man/zhu2019EL.html">zhu2019EL</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">## An instance of class QFeatures containing 62 assays:</span>
<span class="co">##  [1] 1H1a: SingleCellExperiment with 152 rows and 1 columns </span>
<span class="co">##  [2] 1H1b: SingleCellExperiment with 267 rows and 1 columns </span>
<span class="co">##  [3] 1H1c: SingleCellExperiment with 128 rows and 1 columns </span>
<span class="co">##  ...</span>
<span class="co">##  [60] 2N0c: SingleCellExperiment with 61 rows and 1 columns </span>
<span class="co">##  [61] peptides: SingleCellExperiment with 3444 rows and 60 columns </span>
<span class="co">##  [62] proteins: SingleCellExperiment with 840 rows and 60 columns</span></code></pre></div>
</div>
<div id="reproduce-a-published-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#reproduce-a-published-analysis" class="anchor"></a>Reproduce a published analysis</h1>
<p>As a use-case, we will reproduce the analysis of the SCoPE2 data published in <span class="citation">Specht et al. (2021)</span>. SCoPE2 is the first published SCP protocol that has been used to profile thousands of proteins in thousands of single-cells. This is a technical milestone for the field and it opens the door for a fine-grain understanding of biological processes at the protein level.</p>
<p>Along their acquisition protocol, the authors have also provided an <a href="https://github.com/SlavovLab/SCoPE2/blob/8780805f69f7547a2f3b592d3614b649179dfc4e/code/SCoPE2_analysis.R">R script</a> to reproduce their data processing. The code is hard to read and built from scratch. We have used <code>scp</code> to standardize this workflow (<span class="citation">Vanderaa and Gatto (2021)</span>) and reproduce the results using code that can easily be adapted to other data sets. The outline of the workflow is shown below.</p>
<div class="figure">
<img src="figs/SCoPE2_workflow.png" alt="Workflow describing the main steps performed by Specht et al. to process the SCP data." width="2133"><p class="caption">
Workflow describing the main steps performed by Specht et al. to process the SCP data.
</p>
</div>
<div id="import-the-data" class="section level2">
<h2 class="hasAnchor">
<a href="#import-the-data" class="anchor"></a>Import the data</h2>
<p>In a previous section we have shown how to format the feature and sample tables to a <code>Qfeatures</code> object for a small subset of the SCoPE2 data set. The <code>scpdata</code> package provides the <code>Qfeautres</code> object containing the full data set.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="op">(</span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scpdata/man/specht2019v3.html">specht2019v3</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="co">## An instance of class QFeatures containing 179 assays:</span>
<span class="co">##  [1] 190222S_LCA9_X_FP94AA: SingleCellExperiment with 2777 rows and 11 columns </span>
<span class="co">##  [2] 190222S_LCA9_X_FP94AB: SingleCellExperiment with 4348 rows and 11 columns </span>
<span class="co">##  [3] 190222S_LCA9_X_FP94AC: SingleCellExperiment with 4917 rows and 11 columns </span>
<span class="co">##  ...</span>
<span class="co">##  [177] 191110S_LCB7_X_APNOV16plex2_Set_9: SingleCellExperiment with 4934 rows and 16 columns </span>
<span class="co">##  [178] peptides: SingleCellExperiment with 9354 rows and 1490 columns </span>
<span class="co">##  [179] proteins: SingleCellExperiment with 3042 rows and 1490 columns</span></code></pre></div>
<p>The data set provides 177 assays containing the PSM data, but also two assays that contain the peptide and protein data generated by the authors in the original work.</p>
<p>First, to reduce the computational burden we will focus on part of the data. The samples that were acquired in 4 chromatographic batches and we will run this example on the batch called (<code>LCB3</code>). We subset the <code>Qfeatures</code> using the triple-index brackets <code>[feature, sample, assay]</code> and using the LC batch annotation available from the <code>colData</code>. Note also that we use <code>drop = TRUE</code> to drop assays that do not contains samples acquired in <code>LCB7</code> batch.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># TODO Takes too much time...</span>
<span class="co"># scp &lt;- scp[, scp$lcbatch == "LCB3", , drop = TRUE]</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="va">scp</span><span class="op">[</span>, , <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"LCB3"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>
<span class="co">## Warning: 'experiments' dropped; see 'metadata'</span></code></pre></div>
</div>
<div id="psm-quality-control" class="section level2">
<h2 class="hasAnchor">
<a href="#psm-quality-control" class="anchor"></a>PSM quality control</h2>
<div id="filter-out-failed-runs-based-on-psm-content" class="section level3">
<h3 class="hasAnchor">
<a href="#filter-out-failed-runs-based-on-psm-content" class="anchor"></a>Filter out failed runs based on PSM content</h3>
<p>First, only the assays that have sufficient PSMs are kept. The authors keep an assay if it has over 500 PSMs. Before filtering, let’s first look at the distribution of the number of PSMs per assay. Note that we can easily extract the number of rows (here PSMs) and the number of columns (TMT channels) of each assay using the <code>dims</code> function implemented in <code>QFeatures</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">nPSMs</span> <span class="op">&lt;-</span> <span class="fu">dims</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span></code></pre></div>
<p>Let’s have a look at the number of features that were identified in the different runs.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggplot</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">nPSMs</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">nPSMs</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_histogram</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></code></pre></div>
<p><img src="v02-scp_files/figure-html/plot_dims-1.png" width="700"></p>
<p>No MS run failed in the LB3 batch. If some runs had failed, we could have subsetted the data taking advantage of the subsetting method of a <code>QFeatures</code> object.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="va">scp</span><span class="op">[</span>, , <span class="va">nPSMs</span> <span class="op">&gt;</span> <span class="fl">500</span><span class="op">]</span></code></pre></div>
</div>
<div id="filter-out-psms-with-high-false-discovery-rate" class="section level3">
<h3 class="hasAnchor">
<a href="#filter-out-psms-with-high-false-discovery-rate" class="anchor"></a>Filter out PSMs with high false discovery rate</h3>
<p>Next, the SCoPE2 workflow filters PSMs based on the false discovery rate (FDR) for identification. The PSM data were already processed with DART-ID (<span class="citation">(<strong>Chen2019-uc?</strong>)</span>), a python software that updates the confidence in peptide identification using an Bayesian inference approach. DART-ID outputs for every PSM the updated posterior error probability (PEP). Filtering on the PEP is too conservative and it is rather advised to filter based on FDR (<span class="citation">(<strong>Kall2008-hb?</strong>)</span>). To control for the FDR, we need to compute q-values, that correspond to the minimal FDR threshold that would still select the associated feature.</p>
<p>We can use the <code>pep2qvalue</code> function to easily compute q-values from the PEPs computed by MaxQuant or updated by DART-ID. In the SCoPE2 workflow, the features are selected based on the FDR at PSM level and at protein level. The DART-ID PEPs (<code>dart_PEP</code>) as well as the protein information (<code>protein</code>) are automatically retrieved from the <code>rowData</code> of each assay. The function will store the computed q-values back in the <code>rowData</code>. It will be stored under <code>qvalue_psm</code> and <code>qvalue_protein</code> for the PSM and proteins q-values, respectively.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/pep2qvalue.html">pep2qvalue</a></span><span class="op">(</span><span class="va">scp</span>,
                  i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span>,
                  PEP <span class="op">=</span> <span class="st">"dart_PEP"</span>,
                  rowDataName <span class="op">=</span> <span class="st">"qvalue_psm"</span><span class="op">)</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/pep2qvalue.html">pep2qvalue</a></span><span class="op">(</span><span class="va">scp</span>,
                  i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span>,
                  groupBy <span class="op">=</span> <span class="st">"protein"</span>,
                  PEP <span class="op">=</span> <span class="st">"dart_PEP"</span>,
                  rowDataName <span class="op">=</span> <span class="st">"qvalue_protein"</span><span class="op">)</span></code></pre></div>
<p>We can extract the q-values from the <code>rowData</code> of several assays using the <code>rowDataToDF</code> function. It takes the <code>rowData</code> of interest and returns a single <code>DataFrame</code> table with variables of interest. We plot the q-values along with the DART-ID PEPs using <code>ggplot2</code> facets.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/rowDataToDF.html">rowDataToDF</a></span><span class="op">(</span><span class="va">scp</span>,
            i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span>,
            vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"dart_PEP"</span>, <span class="st">"qvalue_psm"</span>, <span class="st">"qvalue_protein"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="va">data.frame</span> <span class="op">%&gt;%</span>
    <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"dart_PEP"</span>, <span class="st">"qvalue_psm"</span>, <span class="st">"qvalue_protein"</span><span class="op">)</span>,
                 names_to <span class="op">=</span> <span class="st">"measure"</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_histogram</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_x_log10</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">facet_grid</span><span class="op">(</span>rows <span class="op">=</span> <span class="fu">vars</span><span class="op">(</span><span class="va">measure</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="v02-scp_files/figure-html/qvalue_plot-1.png" width="700"></p>
<p>We filter the PSMs to control for a 1% PSM and protein FDR. We can perform this on our <code>QFeatures</code> object by using the <code>filterFeatures</code> function. The q-values are directly accessed from the <code>rowData</code> of each assay.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html">filterFeatures</a></span><span class="op">(</span><span class="va">scp</span>,
                      <span class="op">~</span> <span class="va">qvalue_psm</span> <span class="op">&lt;</span> <span class="fl">0.01</span> <span class="op">&amp;</span> <span class="va">qvalue_protein</span> <span class="op">&lt;</span> <span class="fl">0.01</span><span class="op">)</span></code></pre></div>
</div>
<div id="filter-out-contaminants" class="section level3">
<h3 class="hasAnchor">
<a href="#filter-out-contaminants" class="anchor"></a>Filter out contaminants</h3>
<p>We will now remove PSMs that were matched to contaminant proteins (the protein name starts with <code>CON</code>) or to the decoy database (the protein name starts with <code>REV</code>). Again, <code>filterFeatures</code> can directly access the protein names from the <code>rowData</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html">filterFeatures</a></span><span class="op">(</span><span class="va">scp</span>,
                      <span class="op">~</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"REV|CON"</span>, <span class="va">protein</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="filter-out-noisy-spectra" class="section level3">
<h3 class="hasAnchor">
<a href="#filter-out-noisy-spectra" class="anchor"></a>Filter out noisy spectra</h3>
<p>A PIF (parental ion fraction) smaller than 80 % indicates the associated spectra is contaminated by co-isolated peptides and therefore the quantification becomes unreliable. The PIF was computed by MaxQuant and is readily available for filtering.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html">filterFeatures</a></span><span class="op">(</span><span class="va">scp</span>,
                      <span class="op">~</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">PIF</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">PIF</span> <span class="op">&gt;</span> <span class="fl">0.8</span><span class="op">)</span></code></pre></div>
</div>
<div id="filter-out-psms-with-high-sample-to-carrier-ratio" class="section level3">
<h3 class="hasAnchor">
<a href="#filter-out-psms-with-high-sample-to-carrier-ratio" class="anchor"></a>Filter out PSMs with high sample to carrier ratio</h3>
<p>The PSMs are next filtered based on the sample to carrier ratio (SCR), that is the TMT ion intensity of a single-cell sample divided by the TMT ion intensity of the carrier (200 cells) acquired during the same run as the sample. It is expected that the carrier intensities are much higher than the single-cell intensities. We implemented the <code>computeSCR</code> function that computes the SCR for each PSM averaged over all samples of interest in a given assay. A PSM is removed when the mean SCR exceeds 10 %. To perform this, we need to tell the function which columns are the samples of interest and which column is the carrier. The <code>colData</code> of the <code>QFeatures</code> object is used to define this.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">scp</span><span class="op">$</span><span class="va">SampleType</span><span class="op">)</span>
<span class="co">## </span>
<span class="co">##      Blank    Carrier Macrophage   Monocyte  Reference     Unused </span>
<span class="co">##         22         24        192         74         24         48</span></code></pre></div>
<p>In this dataset, <code>SampleType</code> gives the type of sample that is present in each TMT channel. The SCoPE2 protocol includes 5 types of samples:</p>
<ul>
<li>The carrier channels (<code>Carrier</code>) contain 200 cell equivalents and are meant to boost the peptide identification rate.</li>
<li>The normalization channels (<code>Reference</code>) contain 5 cell equivalents and are used to partially correct for between-run variation.</li>
<li>The unused channels (<code>Unused</code>) are channels that are left empty due to isotopic cross-contamination by the carrier channel.</li>
<li>The blank channels (<code>Blank</code>) contain samples that do not contain any cell but are processed as single-cell samples.</li>
<li>The single-cell sample channels contain the single-cell samples of interest (<code>Macrophage</code> or <code>Monocyte</code>).</li>
</ul>
<p>The <code>computeSCR</code> function expects the user to provide a pattern (following regular expression syntax) that uniquely identifies a carrier channel in each run and the samples or blanks. The function will store the mean SCR of each feature in the <code>rowData</code> of each assay.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/computeSCR.html">computeSCR</a></span><span class="op">(</span><span class="va">scp</span>,
                  i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span>,
                  colDataCol <span class="op">=</span> <span class="st">"SampleType"</span>,
                  carrierPattern <span class="op">=</span> <span class="st">"Carrier"</span>,
                  samplePattern <span class="op">=</span> <span class="fl">4</span><span class="op">:</span><span class="fl">16</span><span class="op">)</span>
<span class="co">## Warning in computeSCR(scp, i = names(scp), colDataCol = "SampleType",</span>
<span class="co">## carrierPattern = "Carrier", : The pattern is numeric. This is only allowed for</span>
<span class="co">## replicating the SCoPE2 analysis and will later get defunct.</span></code></pre></div>
<p>In this step, we supplied numeric entries as <code>samplePattern</code> instead of a regular expression pattern. This is to replicate the SCoPE2 analysis. We refrain from using a numeric pattern (as indicated by the warning above) because this can lead to unnoticed artefacts. In fact, the empty channels of the TMT-16 runs but not TMT-11 channels are here also used to compute the average SCR which could have been avoided when using a character pattern.</p>
<p>Before applying the filter, we plot the distribution of the mean SCR.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/rowDataToDF.html">rowDataToDF</a></span><span class="op">(</span><span class="va">scp</span>,
            i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span>,
            vars <span class="op">=</span> <span class="st">"MeanSCR"</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="va">data.frame</span> <span class="op">%&gt;%</span>
    <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">MeanSCR</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_histogram</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">200</span>, <span class="fl">0.1</span><span class="op">)</span>,
               lty <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">1</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_x_log10</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="v02-scp_files/figure-html/meanSCR_plot-1.png" width="700"></p>
<p>A great majority of the PSMs have a mean SCR that is lower than 10%, as expected. Interestingly, the mode of the distribution is located close to 1%. This is expected since every sample channel contains a single-cell and the carrier contains 200 cells leading to an expected ratio of 0.5% (dashed line). We remove the PSMs for which the mean SCR exceeds the 10% threshold.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html">filterFeatures</a></span><span class="op">(</span><span class="va">scp</span>,
                      <span class="op">~</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">MeanSCR</span><span class="op">)</span> <span class="op">&amp;</span>
                          <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/is.finite.html">is.infinite</a></span><span class="op">(</span><span class="va">MeanSCR</span><span class="op">)</span> <span class="op">&amp;</span>
                          <span class="va">MeanSCR</span> <span class="op">&lt;</span> <span class="fl">0.1</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="normalization-to-reference" class="section level2">
<h2 class="hasAnchor">
<a href="#normalization-to-reference" class="anchor"></a>Normalization to reference</h2>
<p>In order to partially correct for between-run variation, SCoPE2 computes relative reporter ion intensities. This means that intensities measured for single-cells are divided by the reference channel (5-cell equivalents). We use the <code>divideByReference</code> function that divides channels of interest by the reference channel. Similarly to <code>computeSCR</code>, we can point to the samples and the reference columns in each assay using the annotation contained in the <code>colData</code>. We will here divide all columns (using the regular expression wildcard <code>.</code>) by the reference channel (<code>Reference</code>). Notice that when taking all samples we also include the reference channel itself. Hence, from now on, the reference channels will contain only ones.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/divideByReference.html">divideByReference</a></span><span class="op">(</span><span class="va">scp</span>,
                         i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span>,
                         colDataCol <span class="op">=</span> <span class="st">"SampleType"</span>,
                         samplePattern <span class="op">=</span> <span class="st">"."</span>,
                         refPattern <span class="op">=</span> <span class="st">"Reference"</span><span class="op">)</span></code></pre></div>
</div>
<div id="psm-to-peptide-aggregation" class="section level2">
<h2 class="hasAnchor">
<a href="#psm-to-peptide-aggregation" class="anchor"></a>PSM to peptide aggregation</h2>
<p>Now that the PSM assays are processed, we can aggregate them to peptides. This is performed using the <code>aggregateFeaturesOverAssays</code> function. This is a wrapper function in <code>scp</code> that sequentially calls the <code>aggregateFeatures</code> from the <code>QFeatures</code> package over the different assays. For each assay, the function aggregates several PSMs into a unique peptide given an aggregating variable in the <code>rowData</code> (peptide sequence) and a user-supplied aggregating function (the median for instance). Regarding the aggregating function, the SCoPE2 analysis removes duplicated peptide sequences per run by taking the first non-missing value. While better alternatives are documented in <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-aggregate.html">QFeatures::aggregateFeatures</a></code>, we still use this approach for the sake of replication and for illustrating that custom functions can be applied.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">remove.duplicates</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">x</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">xx</span><span class="op">)</span> <span class="va">xx</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">xx</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">)</span></code></pre></div>
<p>The aggregated peptide assays must be given a name. We here used the original names with <code>peptides_</code> at the start.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">peptideAssays</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"peptides_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>We now have all the required information to aggregate the PSMs in the different batches to peptides.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/aggregateFeaturesOverAssays.html">aggregateFeaturesOverAssays</a></span><span class="op">(</span><span class="va">scp</span>,
                                   i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span>,
                                   fcol <span class="op">=</span> <span class="st">"peptide"</span>,
                                   name <span class="op">=</span> <span class="va">peptideAssays</span>,
                                   fun <span class="op">=</span> <span class="va">remove.duplicates</span><span class="op">)</span></code></pre></div>
<p>Under the hood, the <code>QFeatures</code> architecture preserves the relationship between the aggregated assays. See <code><a href="https://rdrr.io/pkg/QFeatures/man/AssayLinks.html">?AssayLinks</a></code> for more information on relationships between assays. Notice that <code>aggregateFeaturesOverAssays</code> created as many new assays as the number of supplied assays.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span>
<span class="co">## An instance of class QFeatures containing 48 assays:</span>
<span class="co">##  [1] 190913S_LCB3_X_16plex_Set_12: SingleCellExperiment with 1624 rows and 16 columns </span>
<span class="co">##  [2] 190913S_LCB3_X_16plex_Set_17_reinject2: SingleCellExperiment with 1484 rows and 16 columns </span>
<span class="co">##  [3] 190914S_LCB3_X_16plex_Set_1: SingleCellExperiment with 1415 rows and 16 columns </span>
<span class="co">##  ...</span>
<span class="co">##  [46] peptides_190914S_LCB3_X_16plex_Set_7: SingleCellExperiment with 1775 rows and 16 columns </span>
<span class="co">##  [47] peptides_190914S_LCB3_X_16plex_Set_8: SingleCellExperiment with 1350 rows and 16 columns </span>
<span class="co">##  [48] peptides_190914S_LCB3_X_16plex_Set_9: SingleCellExperiment with 1944 rows and 16 columns</span></code></pre></div>
</div>
<div id="join-the-scope2-sets-in-one-assay" class="section level2">
<h2 class="hasAnchor">
<a href="#join-the-scope2-sets-in-one-assay" class="anchor"></a>Join the SCoPE2 sets in one assay</h2>
<p>Up to now, we kept the data belonging to each MS run in separate assays. We now combine all batches into a single assay. This can easily be done using the <code>joinAssays</code> function from the <code>QFeatures</code> package.</p>
<div id="consensus-mapping-of-peptides-to-proteins" class="section level3">
<h3 class="hasAnchor">
<a href="#consensus-mapping-of-peptides-to-proteins" class="anchor"></a>Consensus mapping of peptides to proteins</h3>
<p>We need to account for an issue in the data. <code>joinAssays</code> will only keep the metadata variables that have the same value between matching rows. However, some peptide sequences map to one protein in one run and to another protein in another run. Hence, the protein sequence is not constant for all peptides and is removed during joining. It is important we keep the protein sequence in the <code>rowData</code> since we will later need it to aggregate peptides to proteins. To avoid this issue, we replace the problematic peptides to protein mappings through a majority vote.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/rowDataToDF.html">rowDataToDF</a></span><span class="op">(</span>i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">24</span><span class="op">]</span>,
                vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"peptide"</span>, <span class="st">"protein"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="va">data.frame</span> <span class="op">%&gt;%</span>
    <span class="fu">group_by</span><span class="op">(</span><span class="va">peptide</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co">## The majority vote happens here</span>
    <span class="fu">mutate</span><span class="op">(</span>protein <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">protein</span><span class="op">)</span>,
                                decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">select</span><span class="op">(</span><span class="va">peptide</span>, <span class="va">protein</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">peptide</span>, <span class="va">protein</span><span class="op">)</span><span class="op">)</span> <span class="op">-&gt;</span>
    <span class="va">ppMap</span>
<span class="va">consensus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">peptideAssays</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">ppMap</span><span class="op">$</span><span class="va">protein</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">scp</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">peptide</span>, <span class="va">ppMap</span><span class="op">$</span><span class="va">peptide</span><span class="op">)</span><span class="op">]</span>
<span class="op">}</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">consensus</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">peptideAssays</span>
<span class="co">## TODO put this in QFeatures</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">replaceRowDataCols</a></span><span class="op">(</span><span class="va">scp</span>, i <span class="op">=</span> <span class="va">peptideAssays</span>, <span class="st">"protein"</span>, <span class="va">consensus</span><span class="op">)</span></code></pre></div>
<p>This code chunk is hard to read, but this issue should represent a special case and we therefore don’t consider this step worth standardizing. Some peptides are mapped to different proteins because the TMT-11 and TMT-16 data sets were analysed separately in MaxQuant. Razor peptides, <em>i.e.</em> peptides that are found in several protein groups, are assigned to the protein group with most peptides (<span class="citation">(<strong>Tyanova2016-yj?</strong>)</span>,) and this can vary from one analysis to another. Therefore, inconsistent peptide mapping should not occur when all runs are preprocessed at once. We will see later that this issue does not impair the replication of the SCoPE2 analysis, but it recalls that mapping of shared peptides to proteins is not a trivial task.</p>
</div>
<div id="cleaning-missing-data" class="section level3">
<h3 class="hasAnchor">
<a href="#cleaning-missing-data" class="anchor"></a>Cleaning missing data</h3>
<p>Another important step before we join the assays is to replace zero and infinite values by <code>NA</code>s. The zeros can be biological zeros or technical zeros and differentiating between the two types is a difficult task, they are therefore better considered as missing. The infinite values arose during the normalization by the reference because the channel values are divide by a zero from the reference channel. This artefact could easily be avoided if we had replace the zeros by <code>NA</code>s at the beginning of the workflow, what we strongly recommend for future analyses.</p>
<p>The <code>infIsNA</code> and the <code>zeroIsNA</code> functions automatically detect infinite and zero values, respectively, and replace them with <code>NA</code>s. Those two functions are provided by the <code>QFeatures</code> package.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html">infIsNA</a></span><span class="op">(</span><span class="va">scp</span>, i <span class="op">=</span> <span class="va">peptideAssays</span><span class="op">)</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html">zeroIsNA</a></span><span class="op">(</span><span class="va">scp</span>, i <span class="op">=</span> <span class="va">peptideAssays</span><span class="op">)</span></code></pre></div>
</div>
<div id="join-assays" class="section level3">
<h3 class="hasAnchor">
<a href="#join-assays" class="anchor"></a>Join assays</h3>
<p>Now that the peptides are correctly matched to proteins and missing values are correctly formatted, we can join the assays.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/joinAssays.html">joinAssays</a></span><span class="op">(</span><span class="va">scp</span>,
                  i <span class="op">=</span> <span class="va">peptideAssays</span>,
                  name <span class="op">=</span> <span class="st">"peptides"</span><span class="op">)</span></code></pre></div>
<p><code>joinAssays</code> has created a new assay called <code>peptides</code> that combines the previously aggregated peptide assays.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span>
<span class="co">## An instance of class QFeatures containing 49 assays:</span>
<span class="co">##  [1] 190913S_LCB3_X_16plex_Set_12: SingleCellExperiment with 1624 rows and 16 columns </span>
<span class="co">##  [2] 190913S_LCB3_X_16plex_Set_17_reinject2: SingleCellExperiment with 1484 rows and 16 columns </span>
<span class="co">##  [3] 190914S_LCB3_X_16plex_Set_1: SingleCellExperiment with 1415 rows and 16 columns </span>
<span class="co">##  ...</span>
<span class="co">##  [47] peptides_190914S_LCB3_X_16plex_Set_8: SingleCellExperiment with 1350 rows and 16 columns </span>
<span class="co">##  [48] peptides_190914S_LCB3_X_16plex_Set_9: SingleCellExperiment with 1944 rows and 16 columns </span>
<span class="co">##  [49] peptides: SingleCellExperiment with 4368 rows and 384 columns</span></code></pre></div>
</div>
</div>
<div id="single-cell-quality-control" class="section level2">
<h2 class="hasAnchor">
<a href="#single-cell-quality-control" class="anchor"></a>Single-cell quality control</h2>
<p>The SCoPE2 script proceeds with filtering the single-cells. The filtering is mainly based on the median coefficient of variation (CV) per cell. The median CV measures the consistency of quantification for a group of peptides that belong to a protein. We remove cells that exhibit high median CV over the different proteins. We compute the median CV per cell using the <code>medianCVperCell</code> function from the <code>scp</code> package. The function takes the protein information from the <code>rowData</code> of the assays that will tell how to group the features (peptides) when computing the CV. Note that we supply the peptide assays before joining in a single assays (<code>i = peptideAssays</code>). This is because SCoPE2 performs a custom normalization (<code>norm = "SCoPE2"</code>). Each row in an assay is normalized by a scaling factor. This scaling factor is the row mean after dividing the columns by the median. The authors retained CVs that are computed using at least 6 peptides (<code>nobs = 6</code>). See the methods section in <span class="citation">Specht et al. (2021)</span> for more information.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/medianCVperCell.html">medianCVperCell</a></span><span class="op">(</span><span class="va">scp</span>,
                       i <span class="op">=</span> <span class="va">peptideAssays</span>,
                       groupBy <span class="op">=</span> <span class="st">"protein"</span>,
                       nobs <span class="op">=</span> <span class="fl">6</span>,
                       na.rm <span class="op">=</span> <span class="cn">TRUE</span>,
                       colDataName <span class="op">=</span> <span class="st">"MedianCV"</span>,
                       norm <span class="op">=</span> <span class="st">"SCoPE2"</span><span class="op">)</span></code></pre></div>
<p>The computed CVs are stored in the <code>colData</code>. We can now filter cells that have reliable quantifications. The blank samples are not expected to have reliable quantifications and hence can be used to estimate a null distribution of the CV. This distribution helps defining a threshold that filters out single-cells that contain noisy quantification.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="va">data.frame</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">SampleType</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Macrophage"</span>,  <span class="st">"Monocyte"</span>, <span class="st">"Blank"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">MedianCV</span>,
               fill <span class="op">=</span> <span class="va">SampleType</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_histogram</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.365</span><span class="op">)</span></code></pre></div>
<p><img src="v02-scp_files/figure-html/medianCV_plot-1.png" width="700"></p>
<p>We can see that the protein quantification for single-cells are much more consistent within single-cell channels than within blank channels. A threshold of 0.365 best separates single-cells from empty channels.</p>
<p>We keep the cells that pass the median CV threshold. Furthermore, we keep macrophages and monocytes as those represent the samples of interest. We can extract the sample names that pass the CV and sample type filters using <code>tidyverse</code> manipulations.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="va">data.frame</span> <span class="op">%&gt;%</span>
    <span class="fu">rownames_to_column</span><span class="op">(</span><span class="st">"cells"</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">MedianCV</span> <span class="op">&lt;</span> <span class="fl">0.365</span>,
           <span class="va">SampleType</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Macrophage"</span>, <span class="st">"Monocyte"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">pull</span><span class="op">(</span><span class="va">cells</span><span class="op">)</span> <span class="op">-&gt;</span>
    <span class="va">keep</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">keep</span><span class="op">)</span>
<span class="co">## [1] "190913S_LCB3_X_16plex_Set_12_RI5"  "190913S_LCB3_X_16plex_Set_12_RI6" </span>
<span class="co">## [3] "190913S_LCB3_X_16plex_Set_12_RI7"  "190913S_LCB3_X_16plex_Set_12_RI8" </span>
<span class="co">## [5] "190913S_LCB3_X_16plex_Set_12_RI9"  "190913S_LCB3_X_16plex_Set_12_RI10"</span></code></pre></div>
<p>We have already demonstrated how to remove assays from the dataset, but we can apply a similar subsetting for the samples. Recall that the second entry in the <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> method is for subsetting sample names.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="va">scp</span><span class="op">[</span>, <span class="va">keep</span>, <span class="op">]</span>
<span class="va">scp</span>
<span class="co">## An instance of class QFeatures containing 49 assays:</span>
<span class="co">##  [1] 190913S_LCB3_X_16plex_Set_12: SingleCellExperiment with 1624 rows and 12 columns </span>
<span class="co">##  [2] 190913S_LCB3_X_16plex_Set_17_reinject2: SingleCellExperiment with 1484 rows and 10 columns </span>
<span class="co">##  [3] 190914S_LCB3_X_16plex_Set_1: SingleCellExperiment with 1415 rows and 11 columns </span>
<span class="co">##  ...</span>
<span class="co">##  [47] peptides_190914S_LCB3_X_16plex_Set_8: SingleCellExperiment with 1350 rows and 11 columns </span>
<span class="co">##  [48] peptides_190914S_LCB3_X_16plex_Set_9: SingleCellExperiment with 1944 rows and 11 columns </span>
<span class="co">##  [49] peptides: SingleCellExperiment with 4368 rows and 265 columns</span></code></pre></div>
</div>
<div id="normalization" class="section level2">
<h2 class="hasAnchor">
<a href="#normalization" class="anchor"></a>Normalization</h2>
<p>The columns (samples) of the peptide data are first normalized by dividing the relative intensities by the median relative intensities. Then, the rows (peptides) are normalized by dividing the relative intensities by the mean relative intensities. The first normalization is available from the <code>normalize</code> function and accessible under the <code>div.median</code> method. The second is not available from <code>normalize</code>, but is easily performed using the <code>sweep</code> function from the <code>QFeatures</code> package that is inspired from the <code><a href="https://rdrr.io/r/base/sweep.html">base::sweep</a></code> function.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Scale column with median</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/normalizeSCP.html">normalizeSCP</a></span><span class="op">(</span><span class="va">scp</span>,
                    i <span class="op">=</span> <span class="st">"peptides"</span>,
                    method <span class="op">=</span> <span class="st">"div.median"</span>,
                    name <span class="op">=</span> <span class="st">"peptides_norm1"</span><span class="op">)</span>
<span class="co">## Scale rows with mean</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html">sweep</a></span><span class="op">(</span><span class="va">scp</span>,
             i <span class="op">=</span> <span class="st">"peptides_norm1"</span>, MARGIN <span class="op">=</span> <span class="fl">1</span>,
             FUN <span class="op">=</span> <span class="st">"/"</span>,
             STATS <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">scp</span><span class="op">[[</span><span class="st">"peptides_norm1"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,
                              na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,
             name <span class="op">=</span> <span class="st">"peptides_norm2"</span><span class="op">)</span></code></pre></div>
<p>Each normalization step is stored in a separate assay. An important aspect to note here is that</p>
</div>
<div id="filter-highly-missing-peptides" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-highly-missing-peptides" class="anchor"></a>Filter highly missing peptides</h2>
<p>Peptides that contain many missing values are not informative. Therefore, we remove those with more than 99 % missing data. This is done using the <code>filterNA</code> function from <code>QFeatures</code>.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html">filterNA</a></span><span class="op">(</span><span class="va">scp</span>,
                i <span class="op">=</span> <span class="st">"peptides_norm2"</span>,
                pNA <span class="op">=</span> <span class="fl">0.99</span><span class="op">)</span></code></pre></div>
</div>
<div id="log-transformation" class="section level2">
<h2 class="hasAnchor">
<a href="#log-transformation" class="anchor"></a>Log-transformation</h2>
<p>The last processing step of the peptide data before aggregating to proteins is to log-transform the data. SCoPE2 uses a base 2 log-transformation and this is implemented in <code>logTransform</code> in <code>QFeatures</code>.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html">logTransform</a></span><span class="op">(</span><span class="va">scp</span>,
                    base <span class="op">=</span> <span class="fl">2</span>,
                    i <span class="op">=</span> <span class="st">"peptides_norm2"</span>,
                    name <span class="op">=</span> <span class="st">"peptides_log"</span><span class="op">)</span></code></pre></div>
</div>
<div id="peptide-to-protein-aggregation" class="section level2">
<h2 class="hasAnchor">
<a href="#peptide-to-protein-aggregation" class="anchor"></a>Peptide to protein aggregation</h2>
<p>Similarly to aggregating PSM data to peptide data, we can aggregate peptide data to protein data using the <code>aggregateFeatures</code> function. Note that we here use the median as a summarizing function.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-aggregate.html">aggregateFeatures</a></span><span class="op">(</span><span class="va">scp</span>,
                         i <span class="op">=</span> <span class="st">"peptides_log"</span>,
                         name <span class="op">=</span> <span class="st">"proteins"</span>,
                         fcol <span class="op">=</span> <span class="st">"protein"</span>,
                         fun <span class="op">=</span> <span class="fu">matrixStats</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/matrixStats/man/rowMedians.html">colMedians</a></span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div id="normalization-1" class="section level2">
<h2 class="hasAnchor">
<a href="#normalization-1" class="anchor"></a>Normalization</h2>
<p>Normalization is performed similarly to peptide normalization. We use the same functions, but since the data were log-transformed at the peptide level, we subtract/center by the statistic (median or mean) instead of dividing.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Center columns with median</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/normalizeSCP.html">normalizeSCP</a></span><span class="op">(</span><span class="va">scp</span>,
                    i <span class="op">=</span> <span class="st">"proteins"</span>,
                    method <span class="op">=</span> <span class="st">"center.median"</span>,
                    name <span class="op">=</span> <span class="st">"proteins_norm1"</span><span class="op">)</span>
<span class="co">## Center rows with mean</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html">sweep</a></span><span class="op">(</span><span class="va">scp</span>,
             i <span class="op">=</span> <span class="st">"proteins_norm1"</span>,
             MARGIN <span class="op">=</span> <span class="fl">1</span>,
             FUN <span class="op">=</span> <span class="st">"-"</span>,
             STATS <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">scp</span><span class="op">[[</span><span class="st">"proteins_norm1"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,
                              na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,
             name <span class="op">=</span> <span class="st">"proteins_norm2"</span><span class="op">)</span></code></pre></div>
</div>
<div id="imputation" class="section level2">
<h2 class="hasAnchor">
<a href="#imputation" class="anchor"></a>Imputation</h2>
<p>The protein data contains a lot of missing values. The graph below shows the distribution of the proportion missingness in cells. Cells contain on average over 75 % missing values!</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">longFormat</a></span><span class="op">(</span><span class="va">scp</span><span class="op">[</span>, , <span class="st">"proteins_norm2"</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="va">data.frame</span> <span class="op">%&gt;%</span>
    <span class="fu">group_by</span><span class="op">(</span><span class="va">colname</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">summarize</span><span class="op">(</span>missingness <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">missingness</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_histogram</span><span class="op">(</span><span class="op">)</span>
<span class="co">## Warning: 'experiments' dropped; see 'metadata'</span></code></pre></div>
<p><img src="v02-scp_files/figure-html/missing_plot-1.png" width="700"></p>
<p>The missing data is imputed using K nearest neighbors. The SCoPE2 script runs KNN with k = 3. We made a wrapper around the author’s code to apply imputation to our <code>QFeatures</code> object.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu">imputeKnnSCoPE2</span><span class="op">(</span><span class="va">scp</span>,
                       i <span class="op">=</span> <span class="st">"proteins_norm2"</span>,
                       name <span class="op">=</span> <span class="st">"proteins_impd"</span>,
                       k <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><code>QFeatures</code> provides the <code>impute</code> function that serves as an interface to different imputation algorithms among which the KNN algorithm from <code>impute::impute.knn</code>. However, the KNN implementation in SCoPE2 and in <code>impute.knn</code> are different. SCoPE2 performs KNN imputation in the sample space, meaning that data from neighbouring cells are used to impute the central cell, whereas <code>impute::impute.knn</code> performs KNN imputation in the feature space, meaning that data from neighbouring features are used to impute the missing values from the central features. We provide the code for KNN imputation with <code>QFeatures</code> but do not run in order to reproduce the SCoPE2 analysis.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/impute.html">impute</a></span><span class="op">(</span><span class="va">scp</span>,
              i <span class="op">=</span> <span class="st">"proteins_norm2"</span>,
              method <span class="op">=</span> <span class="st">"knn"</span>,
              k <span class="op">=</span> <span class="fl">3</span>, rowmax <span class="op">=</span> <span class="fl">1</span>, colmax<span class="op">=</span> <span class="fl">1</span>,
              maxp <span class="op">=</span> <span class="cn">Inf</span>, rng.seed <span class="op">=</span> <span class="fl">1234</span><span class="op">)</span></code></pre></div>
</div>
<div id="batch-correction" class="section level2">
<h2 class="hasAnchor">
<a href="#batch-correction" class="anchor"></a>Batch correction</h2>
<p>The final step is to model the remaining batch effects and correct for it. The data were acquired as a series of MS runs. Recall we had 177 assays at the beginning of the workflow. Each MS run can be subjected to technical perturbations that lead to differences in the data. This must be accounted for to avoid attributing biological effects to technical effects. The <code>ComBat</code> algorithm (<span class="citation">(<strong>Johnson2007-nc?</strong>)</span>) is used in the SCoPE2 script to correct for those batch effects. <code>ComBat</code> is part of the <code>sva</code> package. It requires a batch variable, in this case the LC-MS/MS run, and adjusts for batch effects, while protecting variables of interest, the sample type in this case. All the information is contained in the <code>colData</code> of the <code>QFeatures</code> object. We first extract the assays with the associated <code>colData</code>.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">getWithColData</span><span class="op">(</span><span class="va">scp</span>, <span class="st">"proteins_impd"</span><span class="op">)</span>
<span class="co">## Warning: 'experiments' dropped; see 'metadata'</span></code></pre></div>
<p>We next store the batch variable and create the design matrix. We then perform the batch correction and overwrite the data matrix. Recall the data matrix can be accessed using the <code>assay</code> function. Note also that we here use the <code>ComBatv3.34</code> function that is provided in this package. This function was taken from an older version of the <code>sva</code> package (version <code>3.34.0</code>). This is to replicate the SCoPE2 results. More recent versions will avoid batch correction for proteins that show no variance within at least one batch and this occurs for a significant proportion of the proteins (artefact of imputation).</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">batch</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">Set</span>
<span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">SampleType</span>, data <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>
<span class="fu">assay</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">ComBatv3.34</span><span class="op">(</span>dat <span class="op">=</span> <span class="fu">assay</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>,
                          batch <span class="op">=</span> <span class="va">batch</span>,
                          mod <span class="op">=</span> <span class="va">model</span><span class="op">)</span></code></pre></div>
<p>Finally, we add the batch corrected assay to the <code>QFeatures</code> object and create the feature links.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html">addAssay</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">sce</span>,
             name <span class="op">=</span> <span class="st">"proteins_batchC"</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/AssayLinks.html">addAssayLinkOneToOne</a></span><span class="op">(</span>from <span class="op">=</span> <span class="st">"proteins_impd"</span>,
                         to <span class="op">=</span> <span class="st">"proteins_batchC"</span><span class="op">)</span> <span class="op">-&gt;</span>
    <span class="va">scp</span></code></pre></div>
</div>
<div id="normalization-2" class="section level2">
<h2 class="hasAnchor">
<a href="#normalization-2" class="anchor"></a>Normalization</h2>
<p>Before exporting the data as <code>Proteins-processed.csv</code>, the authors performed an additional normalization step.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Center columns with median</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://UCLouvain-CBIO.github.io/scp/reference/normalizeSCP.html">normalizeSCP</a></span><span class="op">(</span><span class="va">scp</span>,
                    i <span class="op">=</span> <span class="st">"proteins_batchC"</span>,
                    method <span class="op">=</span> <span class="st">"center.median"</span>,
                    name <span class="op">=</span> <span class="st">"proteins_batchC_norm1"</span><span class="op">)</span>
<span class="co">## Center rows with mean</span>
<span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html">sweep</a></span><span class="op">(</span><span class="va">scp</span>,
             i <span class="op">=</span> <span class="st">"proteins_batchC_norm1"</span>,
             MARGIN <span class="op">=</span> <span class="fl">1</span>,
             FUN <span class="op">=</span> <span class="st">"-"</span>,
             STATS <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">scp</span><span class="op">[[</span><span class="st">"proteins_batchC_norm1"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,
                              na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,
             name <span class="op">=</span> <span class="st">"proteins_scp"</span><span class="op">)</span></code></pre></div>
<p>We named the last assay <code>proteins_scp</code>. We will contrast this data matrix to <code>proteins_SCoPE2</code> later in the next section to assess the accuracy of the replication.</p>
</div>
</div>
<div id="downstream-analyses" class="section level1">
<h1 class="hasAnchor">
<a href="#downstream-analyses" class="anchor"></a>Downstream analyses</h1>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">getWithColData</span><span class="op">(</span><span class="va">scp</span>, <span class="st">"proteins_scp"</span><span class="op">)</span>
<span class="co">## Warning: 'experiments' dropped; see 'metadata'</span>
<span class="co">## Warning: Ignoring redundant column names in 'colData(x)':</span></code></pre></div>
<div id="dimension-reduction" class="section level2">
<h2 class="hasAnchor">
<a href="#dimension-reduction" class="anchor"></a>Dimension reduction</h2>
<div id="pca" class="section level3">
<h3 class="hasAnchor">
<a href="#pca" class="anchor"></a>PCA</h3>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span>
<span class="co">## Perform PCA, see ?runPCA for more info about arguments</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runPCA.html">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, 
              ncomponents <span class="op">=</span> <span class="fl">50</span>,
              ntop <span class="op">=</span> <span class="cn">Inf</span>,
              scale <span class="op">=</span> <span class="cn">TRUE</span>,
              exprs_values <span class="op">=</span> <span class="fl">1</span>,
              name <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span> 
<span class="co">## Plotting is performed in a single line of code</span>
<span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotPCA</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"SampleType"</span><span class="op">)</span></code></pre></div>
<p><img src="v02-scp_files/figure-html/PCA_scater-1.png" width="700"></p>
</div>
<div id="tsne" class="section level3">
<h3 class="hasAnchor">
<a href="#tsne" class="anchor"></a>tSNE</h3>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html">runTSNE</a></span><span class="op">(</span><span class="va">sce</span>, 
               ncomponents <span class="op">=</span> <span class="fl">2</span>,
               ntop <span class="op">=</span> <span class="cn">Inf</span>,
               scale <span class="op">=</span> <span class="cn">TRUE</span>,
               exprs_values <span class="op">=</span> <span class="fl">1</span>,
               name <span class="op">=</span> <span class="st">"TSNE"</span><span class="op">)</span> 
<span class="co">## Plotting is performed in a single line of code</span>
<span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotTSNE</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"lcbatch"</span><span class="op">)</span> </code></pre></div>
<p><img src="v02-scp_files/figure-html/unnamed-chunk-28-1.png" width="700"></p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotTSNE</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"SampleType"</span><span class="op">)</span></code></pre></div>
<p><img src="v02-scp_files/figure-html/unnamed-chunk-28-2.png" width="700"></p>
</div>
</div>
<div id="cluster-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#cluster-analysis" class="anchor"></a>Cluster analysis</h2>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">BiocNeighbors</span><span class="op">)</span>
<span class="va">sce</span><span class="op">$</span><span class="va">Cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/clusterCells.html">clusterCells</a></span><span class="op">(</span><span class="va">sce</span>, 
                    assay.type <span class="op">=</span> <span class="fl">1</span>,
                    BLUSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html">SNNGraphParam</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">10</span>,
                                              type <span class="op">=</span> <span class="st">"rank"</span>,
                                              BNPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocNeighbors/man/KmknnParam.html">KmknnParam</a></span><span class="op">(</span><span class="op">)</span>,
                                              cluster.fun <span class="op">=</span> <span class="st">"walktrap"</span>,
                                              cluster.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotTSNE</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"Cluster"</span><span class="op">)</span> </code></pre></div>
<p><img src="v02-scp_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
</div>
<div id="maker-gene-identification" class="section level2">
<h2 class="hasAnchor">
<a href="#maker-gene-identification" class="anchor"></a>Maker gene identification</h2>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/findMarkers.html">findMarkers</a></span><span class="op">(</span><span class="va">sce</span>, 
                       groups <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">Cluster</span>,
                       assay.type <span class="op">=</span> <span class="fl">1</span>,
                       pval.type <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span>
<span class="va">markers</span><span class="op">$</span><span class="va">`3`</span> <span class="op">%&gt;%</span> 
    <span class="va">data.frame</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu">arrange</span><span class="op">(</span><span class="va">FDR</span><span class="op">)</span>
<span class="co">##             p.value          FDR summary.logFC     logFC.1     logFC.2</span>
<span class="co">## Q9BXJ1 1.933543e-13 3.250286e-10    0.81254710  0.85498545  1.02896745</span>
<span class="co">## Q494X3 2.148701e-08 1.805983e-05   -0.37508464 -0.37508464 -0.40040410</span>
<span class="co">## Q99439 4.755157e-08 2.263554e-05    0.29085513  0.29085513  0.37580655</span>
<span class="co">## Q9NPA2 6.553072e-08 2.263554e-05    0.46835660  0.56275329  0.67042695</span>
<span class="co">## P18089 6.732759e-08 2.263554e-05    0.56234588  0.66290118  0.69492807</span>
<span class="co">## Q5VZF2 5.401449e-07 1.513306e-04    0.39126726  0.43197977  0.39126726</span>
<span class="co">## Q14019 6.770815e-07 1.625963e-04    0.51363935  0.51363935  0.62173810</span>
<span class="co">## B7Z6K7 1.059508e-06 2.226292e-04   -0.14661777 -0.18987904 -0.14661777</span>
<span class="co">## O00487 1.287275e-06 2.404344e-04    0.46500035  0.49573389  0.46500035</span>
<span class="co">## P01563 1.526154e-06 2.565466e-04    0.12942096 -0.19112096 -0.15049154</span>
<span class="co">## Q9BSJ2 3.569442e-06 5.454757e-04    0.13673405  0.15781745  0.13009147</span>
<span class="co">## Q9NVJ2 5.150106e-06 7.214440e-04    0.41745552  0.42332512  0.48305573</span>
<span class="co">## Q9ULD9 5.828259e-06 7.536387e-04    0.55588997  0.52283567  0.53054062</span>
<span class="co">## Q9BXH1 6.700963e-06 8.045942e-04   -0.17058230 -0.27573333 -0.21956420</span>
<span class="co">## Q9H2J1 8.548858e-06 9.580420e-04   -0.14571293 -0.18785798 -0.14311163</span>
<span class="co">## Q7Z5J1 1.394600e-05 1.465201e-03   -0.12680336 -0.18711699 -0.12680336</span>
<span class="co">## P35908 2.125542e-05 2.101786e-03   -0.30099678 -0.39885383 -0.29564549</span>
<span class="co">## Q68DI1 2.548223e-05 2.379757e-03    0.24308406  0.29857456  0.31142801</span>
<span class="co">## Q9H772 2.919661e-05 2.515551e-03    0.49780944  0.49780944  0.73487507</span>
<span class="co">## Q8IXI2 3.224819e-05 2.515551e-03    0.08770299 -0.08687782 -0.09499504</span>
<span class="co">## P62081 3.387701e-05 2.515551e-03    0.27551306  0.37803101  0.27765383</span>
<span class="co">## Q6NUQ4 3.516590e-05 2.515551e-03    0.44211302  0.45128470  0.44211302</span>
<span class="co">## Q9H6R6 3.528246e-05 2.515551e-03    0.16272926  0.30882438  0.16272926</span>
<span class="co">## Q96NY7 3.591506e-05 2.515551e-03   -0.36105499 -0.42241996 -0.32999553</span>
<span class="co">## Q9NYC9 5.399988e-05 3.630952e-03   -0.11560039 -0.23956896 -0.11560039</span>
<span class="co">## O15226 5.923103e-05 3.829514e-03   -0.05551489 -0.07011485 -0.05551489</span>
<span class="co">## O60231 6.545991e-05 4.075485e-03    0.36177085  0.58348903  0.85059829</span>
<span class="co">## P62318 7.671458e-05 4.605615e-03    0.33630155  0.38387362  0.33630155</span>
<span class="co">## Q8WXR4 9.691255e-05 5.617586e-03   -0.10698566 -0.11959029 -0.09674828</span>
<span class="co">## O75121 1.018605e-04 5.707581e-03   -0.35540268  0.61813189  0.64278008</span>
<span class="co">## O75533 1.124684e-04 5.930232e-03    0.11946962  0.29826002  0.11946962</span>
<span class="co">## P62995 1.128896e-04 5.930232e-03   -0.22870279 -0.19836506 -0.27765329</span>
<span class="co">## Q9NVD3 1.241499e-04 6.157485e-03   -0.24383700  0.36615706  0.26789031</span>
<span class="co">## Q96K78 1.245416e-04 6.157485e-03   -0.15919430 -0.49304519 -0.16346515</span>
<span class="co">## Q99814 1.359107e-04 6.527595e-03   -0.21739510  0.31068311  0.22123860</span>
<span class="co">## O95881 1.453188e-04 6.785583e-03   -0.45887221  0.53247570  0.45443238</span>
<span class="co">## P49841 1.511516e-04 6.867185e-03   -0.07334842  0.12744580  0.10981548</span>
<span class="co">## Q6PI26 1.663574e-04 7.359125e-03   -0.08663447  0.14362583  0.10863386</span>
<span class="co">## Q16875 1.932920e-04 8.331379e-03    0.25496314  0.25496314  0.44637984</span>
<span class="co">## Q9H6Q4 2.005161e-04 8.426690e-03   -0.18321900  0.25190284  0.30009851</span>
<span class="co">## Q3MIS6 2.241724e-04 9.165374e-03    0.17887057  0.27914785  0.22662613</span>
<span class="co">## Q9Y4F9 2.289980e-04 9.165374e-03    0.07993566  0.11617421  0.07993566</span>
<span class="co">## P78344 3.092821e-04 1.209077e-02    0.35338331  0.35338331  0.44927167</span>
<span class="co">## P55072 3.388779e-04 1.294667e-02   -0.13395400 -0.24099326 -0.13395400</span>
<span class="co">## Q14896 3.502515e-04 1.308384e-02    0.10625878 -0.11864775 -0.16012510</span>
<span class="co">## P21796 3.775982e-04 1.379875e-02   -0.21773723 -0.28644214 -0.17016033</span>
<span class="co">## Q9UQ90 3.896086e-04 1.393473e-02   -0.26468969  0.24032647  0.34074627</span>
<span class="co">## Q8NHP7 5.412970e-04 1.895667e-02    0.09147512 -0.28363076 -0.17309037</span>
<span class="co">## Q9UNF0 5.891553e-04 2.019578e-02    0.25997058  0.25997058  0.47688273</span>
<span class="co">## Q5JV73 6.007073e-04 2.019578e-02    0.12532491  0.12532491  0.14564237</span>
<span class="co">## A8TX70 6.141000e-04 2.024122e-02    0.10396464  0.10396464  0.25470441</span>
<span class="co">## Q9NVP1 6.892528e-04 2.188460e-02   -0.29168505  0.47896772  0.29261821</span>
<span class="co">## P09972 6.899964e-04 2.188460e-02   -0.27880799  0.57938379  0.54754620</span>
<span class="co">## P30405 7.154816e-04 2.227268e-02    0.47808494  0.72304484  0.75845101</span>
<span class="co">## P22695 7.965668e-04 2.366768e-02   -0.27983299  0.36156204  0.49268344</span>
<span class="co">## P52292 8.111946e-04 2.366768e-02    0.31831724  0.31831724  0.39343020</span>
<span class="co">## O75396 8.138046e-04 2.366768e-02    0.21287615  0.21287615  0.33171428</span>
<span class="co">## Q9GZM3 8.166124e-04 2.366768e-02   -0.22072112 -0.22072112 -0.26392593</span>
<span class="co">## A6NM28 9.947772e-04 2.834272e-02    0.22962801  0.24269043  0.27290259</span>
<span class="co">## Q14147 1.295022e-03 3.585372e-02   -0.07689422  0.10241536 -0.06998846</span>
<span class="co">## Q96PQ6 1.301057e-03 3.585372e-02    0.13016573  0.28953888  0.17206243</span>
<span class="co">## Q6P6C2 1.360054e-03 3.687501e-02    0.14341854  0.14990111  0.14346244</span>
<span class="co">## Q96BT1 1.519426e-03 4.054216e-02   -0.18514287  0.41050892  0.47094545</span>
<span class="co">## Q14739 1.561412e-03 4.101146e-02   -0.13696477 -0.20538086 -0.13696477</span>
<span class="co">## P56192 1.590258e-03 4.112652e-02    0.06117148  0.14104735  0.06117148</span>
<span class="co">## Q9NZI8 1.896258e-03 4.829712e-02    0.10201325  0.22900143  0.10201325</span>
<span class="co">##            logFC.4     logFC.5</span>
<span class="co">## Q9BXJ1  0.86359164  0.81254710</span>
<span class="co">## Q494X3 -0.43550715 -0.48177578</span>
<span class="co">## Q99439 -0.30519508  0.46819840</span>
<span class="co">## Q9NPA2  0.46835660  0.70648456</span>
<span class="co">## P18089  0.56234588  0.72303025</span>
<span class="co">## Q5VZF2 -0.49043982  0.42039538</span>
<span class="co">## Q14019 -1.38920760  0.54289089</span>
<span class="co">## B7Z6K7  0.19411128 -0.17847728</span>
<span class="co">## O00487  0.70121843  0.66651940</span>
<span class="co">## P01563  0.12942096 -0.18723421</span>
<span class="co">## Q9BSJ2  0.13673405  0.15781946</span>
<span class="co">## Q9NVJ2  0.41745552  0.51652684</span>
<span class="co">## Q9ULD9  0.55588997  0.50716600</span>
<span class="co">## Q9BXH1 -0.17058230 -0.32279591</span>
<span class="co">## Q9H2J1 -0.14571293 -0.21109526</span>
<span class="co">## Q7Z5J1 -0.19990057 -0.21216307</span>
<span class="co">## P35908 -0.30099678 -0.58747057</span>
<span class="co">## Q68DI1 -0.28105478  0.24308406</span>
<span class="co">## Q9H772 -0.93759476  0.53870293</span>
<span class="co">## Q8IXI2  0.08770299 -0.11032676</span>
<span class="co">## P62081  0.40622420  0.27551306</span>
<span class="co">## Q6NUQ4 -0.75242406  0.51934134</span>
<span class="co">## Q9H6R6  0.24481462  0.26856567</span>
<span class="co">## Q96NY7 -0.36105499 -0.36894215</span>
<span class="co">## Q9NYC9 -0.35489440 -0.11627907</span>
<span class="co">## O15226  0.07418899 -0.07565645</span>
<span class="co">## O60231  0.36177085  0.77946506</span>
<span class="co">## P62318 -1.00029577  0.35124908</span>
<span class="co">## Q8WXR4 -0.10314411 -0.10698566</span>
<span class="co">## O75121 -0.35540268  0.58485847</span>
<span class="co">## O75533  0.19542402  0.16561427</span>
<span class="co">## P62995 -0.22870279 -0.32397524</span>
<span class="co">## Q9NVD3 -0.24383700  0.38920269</span>
<span class="co">## Q96K78 -0.31976168 -0.15919430</span>
<span class="co">## Q99814 -0.21739510  0.29434813</span>
<span class="co">## O95881 -0.45887221  0.45465436</span>
<span class="co">## P49841 -0.07334842  0.16004964</span>
<span class="co">## Q6PI26 -0.08663447  0.14493303</span>
<span class="co">## Q16875 -0.35799122  0.31908963</span>
<span class="co">## Q9H6Q4 -0.18321900  0.28405764</span>
<span class="co">## Q3MIS6  0.17887057  0.23126025</span>
<span class="co">## Q9Y4F9  0.14371898  0.10656503</span>
<span class="co">## P78344 -0.97472090  0.45709941</span>
<span class="co">## P55072 -0.22362898 -0.16225150</span>
<span class="co">## Q14896  0.10625878 -0.16126292</span>
<span class="co">## P21796 -0.21773723 -0.25992858</span>
<span class="co">## Q9UQ90 -0.26468969  0.35820398</span>
<span class="co">## Q8NHP7  0.09147512 -0.26187927</span>
<span class="co">## Q9UNF0 -0.45562831  0.46919521</span>
<span class="co">## Q5JV73  0.17617242  0.24148729</span>
<span class="co">## A8TX70 -0.14358143  0.18536613</span>
<span class="co">## Q9NVP1 -0.29168505  0.42172349</span>
<span class="co">## P09972 -0.27880799  0.64802465</span>
<span class="co">## P30405  0.47808494  0.79442469</span>
<span class="co">## P22695 -0.27983299  0.37495456</span>
<span class="co">## P52292  0.48873958  0.34508609</span>
<span class="co">## O75396 -0.38855694  0.22951421</span>
<span class="co">## Q9GZM3 -0.55122582 -0.27718033</span>
<span class="co">## A6NM28  0.22962801  0.28640670</span>
<span class="co">## Q14147  0.14088088 -0.07689422</span>
<span class="co">## Q96PQ6 -0.33504887  0.13016573</span>
<span class="co">## Q6P6C2 -0.41955715  0.14341854</span>
<span class="co">## Q96BT1 -0.18514287  0.43475248</span>
<span class="co">## Q14739 -0.18935075 -0.22008741</span>
<span class="co">## P56192  0.17162674  0.08843624</span>
<span class="co">## Q9NZI8  0.21697881  0.29590812</span></code></pre></div>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Kelly2020-xd" class="csl-entry">
Kelly, Ryan T. 2020. <span>“<span>Single-Cell</span> Proteomics: Progress and Prospects.”</span> <em>Mol. Cell. Proteomics</em>, August.
</div>
<div id="ref-Lun2020-cv" class="csl-entry">
Lun, Aaron, and Davide Risso. 2020. <em><span>SingleCellExperiment</span>: <span>S4</span> Classes for Single Cell Data</em>.
</div>
<div id="ref-Specht2021-jm" class="csl-entry">
Specht, Harrison, Edward Emmott, Aleksandra A Petelski, R Gray Huffman, David H Perlman, Marco Serra, Peter Kharchenko, Antonius Koller, and Nikolai Slavov. 2021. <span>“Single-Cell Proteomic and Transcriptomic Analysis of Macrophage Heterogeneity Using <span>SCoPE2</span>.”</span> <em>Genome Biol.</em> 22 (1): 50.
</div>
<div id="ref-Vanderaa2021-xd" class="csl-entry">
Vanderaa, Christophe, and Laurent Gatto. 2021. <span>“Utilizing Scp for the Analysis and Replication of Single-Cell Proteomics Data.”</span> <em>bioRxiv</em>.
</div>
<div id="ref-Zhu2019-ja" class="csl-entry">
Zhu, Ying, Mirko Scheibinger, Daniel Christian Ellwanger, Jocelyn F Krey, Dongseok Choi, Ryan T Kelly, Stefan Heller, and Peter G Barr-Gillespie. 2019. <span>“Single-Cell Proteomics Reveals Changes in Expression During Hair-Cell Development.”</span> <em>Elife</em> 8 (November).
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Laurent Gatto, Christophe Vanderaa.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
